### DISCUSSION TITLE
Suggestion: Add automatic performance calculator to aid reverse engineering

### DISCUSSION THREAD

**Discussion by gustavonmartins** [JSBSim-Team/jsbsim#634] [Ideas]
Hi, all.

**Background:**

I have observed that in FlightGear, some of the input values of the FDM dont have official documents as a direct source. Rather, they are based on hard "reverse engineering", trying to make the aircraft behave as some pilots says it should be, or by tweaking until some  specific behavior from aircraft matches real life expectations.

If real life values were being inputted, there would be never a doubt on them, but as this is a result of hard reverse engineering work, sometimes it can happen that by adjusting some parameter, we are inadvertently getting some undesired side effects somewhere else. 

**Proposal:**
To allow JSBSIM supporting "reverse engineering" by generating automatically some performance parameters which are easy to cross-check with airplanes Pilot Operating Handbook (specially on chapter 5, performance). For instance:

-JSBSIM could automatically generate best glide speed and ratio based on aerodynamic coefficients, so if some input data gives the best L/D of 13, but the real aircraft is known to have it at 20, the developer would automatically know that he did something wrong.

-JSBSIM could automatically generate the maximum aircraft service ceiling based on powerplant power avaiable and required power to mantain level flight

-JSBSIM could automatically generate maximum vx and vy for climb for a ISA atmosphere, so the designer could compare it against the POH

-JSBSIM could automatically generate many other standard performance behavior as is typically found on chapter 5 (performance) of many aircraft's Pilot Operating Handbook.

It would also be good if the designer set some goal values (for instance, 700 fpm max vertical climb on cessna 172 when using 75 knots kias), and jsbsim would perform an automated test to tell if this criteria has been met (inside some tolerance).

**Expected benefit:**

With his feature I believe that many regression bugs on aircraft performance would be caught before they entered into production, and would allow developers to adjust their models to implement pilots feedback more readily and lose the fear of inadvertently damaging their aircraft model.

Please, lets discuss if any idea is unclear :)

Thanks

**Comment by seanmcleod** [comment]
@namoscagnm I'd be interested in working on this.

In a similar vein I've been working on and off for a while on a tool to generate Energy Maneuverability diagrams based on specified aircraft performance, see https://github.com/seanmcleod/EMDiagrams/blob/master/EMDiagrams.ipynb

I've also got an example of trying to match 'tales' of A4 roll performance with some recorded data and based on aerodynamic performance data from a NASA report.

https://github.com/seanmcleod/JupyterFTE/blob/master/A4%20-%20Skyhawk%20Roll%20Performance.ipynb

Lastly in the same repo (Jupyter FTE - Flight Test Engineering) you'll see some examples of trying to calculate level acceleration based on real FTI data recorded in a small aircraft, the Blackshape Prime.

https://github.com/seanmcleod/JupyterFTE

**Reply by gustavonmartins** [reply]
Hi, @seanmcleod. 

I firstly created this issue under an aicraft gitub (for the c172 team of flight gear), and I got feedback that this feature needs to work without requiring the FG's aircraft developer to know programming. I think that this is a quite important restritiction. I started by thinking the ac creator could code, but real life says its too much of an expectation for them. Thus, this had to be in a way that requires no more coding skills than the normal ac developer would have... maybe just creating an xml file with expected results.

You can see the original thread here:
https://github.com/c172p-team/c172p/issues/1402

I also saw how the team gathers pilot input and their concerns over breaking some performance, which is here:
https://github.com/c172p-team/c172p/issues/1364
This is a quite interesting conversation and will give you many insights on how to do something useful for every one. This is the issue that actually gave me the idea about this feature for aiding reverse engineering.

Thanks for your time

**Comment by seanmcleod** [comment]
> -JSBSIM could....

Just to be clear, I don't see this being some integral feature of JSBSim, rather a separate tool either in the JSBSim repo like Aeromatic++ or in a separate repo like JSBSimCommander.

In terms of 2nd link above unfortunately I don't see that as a particularly good example for the sort of tool we're talking about. The whole thread is about more subjective rather than objective observations, i.e. whether someone feels there is "too much" torque induced roll, the feel of the controls etc. I didn't see a specific example of someone pointing to the manufacturer's data saying that under these conditions (power, IAS, altitude etc.) the aircraft needs x amount of aileron or rudder trim from say official test data (a lot of data will only be available in the test reports/data during the test program, as opposed to ending up in the POH).

In general handling qualities is going to be the trickiest.

My idea with the JupyterFTE repo was to come up with a number of auto-pilots that could perform a number of test techniques, e.g. to perform the required test technique for level acceleration, then record the data and run it through the particular FTE analysis to derive the particular performance data to then compare to published performance data.

So other examples would be calculating $V_{MCG}$, $V_{MU}$, $V_{MC}$ ( @agodemar some Latex math in a discussion 😉 ) etc. which also involve a particular test pilot technique to be implemented as an auto pilot and then analysing the data.

Example of a paper implementing a model and comparing it's $V_{MCG}$ values with those published by Boeing.

![image](https://user-images.githubusercontent.com/824856/170862804-3353da90-7f4a-4157-89c5-deaef52b4cb1.png)

**Reply by agodemar** [reply]
> Just to be clear, I don't see this being some integral feature of JSBSim, rather a separate tool either in the JSBSim repo like Aeromatic++ or in a separate repo like JSBSimCommander.

@namoscagnm I agree with @seanmcleod, this could be a side project that we can use to establish a fixed set of procedures to assess aircraft performance that come out from a given FDM.

**Reply by gustavonmartins** [reply]
Hi. I am new in the flight gear/ jsbsim world, so my feedback might need some adaptations to real life, so feel free to create this feature as a separate project.

It's still important that this would be as easy as possible to use to the ac developers and also should be made visible for them, as a way to actively promote the generation of better FDM input data.

@seanmcleod I saw some of your code and it's really interesting. I initially thought of making an auto pilot to test performance. Some cases will be easier because we just have to perform some math on the data ( for instance to find best glide, which is just giving the angle with best L/D) instead of running an auto pilot .

For vx or vy I don't know if there is easy math for that (I had my controls/performance class at my aerospace degree 16 years ago so I might have forgotten it).

About the way of the ac developers to interact with it, I think that some can be done with XML files for standard stuff like vx, vy etc, but if now complex scenarios are needed them maybe python would be a good method.

Another scenario to consider is:
When some special  procedure needs to be performed before testing (for instance, turning on the engines and leaning the mixture) , we might have to:
1) reproduce the procedure in python.
2) use pre existing nasal code from the AC to auto start it
3) force the engine to be running on the FDM, bypassing checklists/required pre conditions.

I'm not really well versed into JSBSIM to know which of the three options are realistic, but I believe that if we have to go on option 2, jsbsim would need some way of letting the nasal code operate it, as many pre existing ac logic for auto start and auto pilot are using nasal files. This capability would have to be built early on, otherwise it would get exponentially more difficult to build it afterwards.

After we have this solved, I think it could proceed to think of the first item that we would make available for those automated performance checks. For this out would be better to get ac developer feedback first.

Thanks for this nice discussion!

**Comment by dany93** [comment]
(Sorry @namoscagnm  for having taken some time to respond)

This tool could be helpful without doubts to check aircraft performance for particular circumstances. Whether aircraft developers will prefer using it or make tests and measurements in flight remains to be seen. Some measurements (such as drag, best glide ratio) are cumbersome to carry out in-flight, but flight tests enable to observe and check other features at the same time.

For instance, the best glide ratio is the result of form drag and induced drag (amplitude dependence vs airspeed for each one). Adjusting form drag and induced drag to get the correct best glide ratio at the correct airspeed will not be enough to ensure the correct drag amplitude at every airspeed. This will still need many tests at other airspeeds. Granted, each individual test will be obtained (numerical values by calculation) more quickly than in the simulator. However, one has to enter the correct parameter set for each situation.

The possible use by aircraft FDM developers (who do not need to be experts in computer language) would depend on its complexity and the more or less limited scope in results. Aircraft FDM developers need to know XML language (easy), XML instructions (not that easy, but unavoidable and can be learnt from FlightGear), and some nasal. Personally, I just have a basic nasal skill and almost no knowledge in C++ (just reading a few isolated lines).
Moreover, this practice is made easier with the AEROMATIC template.

@namoscagnm wrote
>I have observed that in FlightGear, some of the input values of the FDM dont have official documents as a direct source. Rather, they are based on hard "reverse engineering", trying to make the aircraft behave as some pilots says it should be, or by tweaking until some specific behavior from aircraft matches real life expectations.

Not exactly, not only based on hard "reverse engineering".
AEROMATIC is a good start.
Megginson's coefficients (calculated, based on Roskam's computer models) often are a (very useful) good reference.
Reverse engineering is used at the last steps of adjustment.
Wind tunnel coefficients can sometimes be found in the literature, only for a limited set, not always more reliable than the calculated ones.

Another difficulty, partly due to imperfections from the FDM itself, is that aircraft developers will always have to do compromises between the final criteria regarding different results.
The issue of degrading a behavior by adjusting another one will always exist. However, we can expect that this tool will make tests more quick or less cumbersome to carry out.

Obviously, this tool would not be usable for complicated behaviors like felt roughness at controls, stall and spin behavior, how to evaluate the adverse yaw ... Based on which criteria can one say that the stall behavior is realistic?
I think that we will never avoid flight tests, that there are too many criteria (some of them subjective) to be taken into account.

**Comment by seanmcleod** [comment]
> Some cases will be easier because we just have to perform some math on the data ( for instance to find best glide, which is just giving the angle with best L/D) instead of running an auto pilot .

In general I see 3 cases:

1. Offline (i.e. no need to fly) calculations based on some coefficients using known performance formula estimates, e.g. like the A4 roll performance example:

$ \\large p_{ss} = -\\frac{2u_0}{b} \\frac{C_{l \\delta a}}{C_{lp}} \\Delta \\delta_a $

However in general this won't be easy to automate since there is so much flexibility in the way that users can enter their force and moment functions, it's not as if they simply fill in values for some well-known coefficients. 

2. Basic auto-pilot. Really just scripted control inputs like doublets etc. e.g. to excite the short-period and then the software records the resulting data and analyses it to compute the short period frequency, damping ratio etc.

3. PID based auto-pilots. Where the auto-pilot is more complicated than a simple scripted control input, instead it will be given some set-point, e.g. to maintain some IAS while power is increased to for example measure climb performance at different speeds etc.

What I suggest we do is come up with say 5 example performance metrics as a starting point and use them to figure out how we would go about automating the collection of the required performance data.

**Comment by gustavonmartins** [comment]
Hi, I have talked to a university classmate who is now a test engineer at his country's airforce. He has planned and executed many tests with airforce pilots and he recommended two books describe which tests are relavant and how to perform it. It will be probably very useful to for us:

1) Flight Testing of Fixed-Wing Aircraft, by Ralph D Kimberlin
2) Introduction to Aerospace Engineering with a Flight Test Perspective, by Stephen Corda

I am browsing thru the book from Kinberlin on my rate sparte time and its quite inspiring and simple... I think you will also enjoy it a lot, @seanmcleod

**Comment by seanmcleod** [comment]
I'll take a look and see if I can find some excerpts from the books.

Once you've gone through enough examples in the book then as I suggested I think you should propose say 5 specific performance tests and then we use those as a starting point to see how easy/difficult they would be to implement.
